### PR TITLE
SRW-7: Add testing

### LINKED ISSUE #7: Add proper tests for pull requests
## What
Add proper tests for the site to be run for each pull request
## Acceptance
Tests should make sure that building and running the application works

### PR DESCRIPTION
**Pull request by erunks**
## What
_TESTING!_ We finally have testing now added to the repo, so I wholly expect everyone to contribute to testing from now on. This also sets up a `pre-commit` hook and should run the tests in a github workflow, as a final check.

## [Link to Issue](https://github.com/speedrun-website/speedrun.website/issues/7)
Resolves #7 

### Steps for testing
- [x] Github `test` workflow should run and pass

### Images
N/A

### CODE REVIEW

File: .prettierrc [hunk 1]
```diff
@@ -1,5 +1,5 @@
 {
-  "semi": false,
+  "semi": true,
```
Comment (legrandgrand) [inline .prettierrc:2]: Tried to check for prettier and every file had issues. Can yourun prettier --write . so everything is correct from here and husky can take over?

File: package.json [hunk 2]
```diff
@@ -7,14 +7,24 @@
...
   },
   "dependencies": {
     "@nuxtjs/axios": "5.13.6",
     "@nuxtjs/pwa": "3.3.5",
+    "@testing-library/jest-dom": "^5.14.1",
+    "@testing-library/user-event": "^13.2.0",
+    "@testing-library/vue": "^5.8.1",
     "core-js": "3.15.1",
+    "husky": "^7.0.1",
     "nuxt": "2.15.7"
```
Comment (legrandgrand) [inline package.json:28]: My IDE yelled at me because Vue isn't in the dependency list... we should probably add it ^^"

File: jest.setup.js [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+ * so that they can be found by jest without any complications.
+ *
+ * Solution is based on the following example:
+ * https://github.com/nuxt/components/issues/58#issuecomment-709961688
+ */
+const globalComponentPaths = [
+  './components/**/base/**/*.vue',
+  './components/**/core/**/*.vue',
+];
+forEach(globalComponentPaths, (globalComponentPath) => {
```
Comment (legrandgrand) [inline jest.setup.js:20]: I understand using lodash for join and uniq, but forEach is native, so we should use that instead :)

File: jest.setup.js [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+  // glob returns an array of filenames matching the current globalComponentPath pattern
+  forEach(glob.sync(path.join(__dirname, globalComponentPath)), (file) => {
```
Comment (legrandgrand) [inline jest.setup.js:22]: same here

File: jest.setup.js [hunk 5]
```diff
@@ -0,0 +1,39 @@
+import '@testing-library/jest-dom';
+import path from 'path';
+import glob from 'glob';
```
Comment (legrandgrand) [inline jest.setup.js:3]: This dep is imported but it's not in the package.json

File: jest.setup.js [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+];
+forEach(globalComponentPaths, (globalComponentPath) => {
+  // glob returns an array of filenames matching the current globalComponentPath pattern
+  forEach(glob.sync(path.join(__dirname, globalComponentPath)), (file) => {
+    const componentRoutes = file.split('/components/')[1].split('/');
+    componentRoutes[componentRoutes.length - 1] = componentRoutes[
+      componentRoutes.length - 1
+    ].replace('.vue', ''); // remove .vue from the last element
+    componentRoutes[0] = upperFirst(componentRoutes[0]);
+    const nuxtName = join(uniq(componentRoutes), '');
```
Comment (legrandgrand) [inline jest.setup.js:28]: same here: join exists natively

File: components/SiteNavbar/SiteNavbar.test.js [hunk 7]
```diff
@@ -0,0 +1,16 @@
...
+  it('should render without crashing', () => {
+    const { unmount } = stubbedRender(SiteNavbar);
+
+    unmount();
+  });
+
+  it('renders correctly', () => {
+    stubbedRender(SiteNavbar);
+
+    expect(document.querySelector('.site-navbar')).toMatchSnapshot();
```
[the remark this answers is not in the document]
Comment (legrandgrand) [reply components/SiteNavbar/SiteNavbar.test.js:14]:
```
const { container } = stubbedRender(SiteNavbar);

expect(container.firstChild).toMatchSnapshot();
```

File: components/SiteFooter/SiteFooter.test.js [hunk 8]
```diff
@@ -0,0 +1,16 @@
...
+  it('should render without crashing', () => {
+    const { unmount } = stubbedRender(SiteFooter);
+
+    unmount();
+  });
+
+  it('renders correctly', () => {
+    stubbedRender(SiteFooter);
+
+    expect(document.querySelector('.site-footer')).toMatchSnapshot();
```
Comment (legrandgrand) [inline components/SiteFooter/SiteFooter.test.js:14]: Same comment as SiteNavbar

File: components/base/ButtonLink/ButtonLink.test.js [hunk 9]
```diff
@@ -0,0 +1,58 @@
...
+  });
+
+  it('renders correctly', () => {
+    stubbedRender(ButtonLink, {
+      attrs: defaultAttrs,
+      props: defaultProps,
+      slots: { default: 'Button' },
+    });
+
+    expect(document.querySelector('.custom-link')).toMatchSnapshot();
```
Comment (legrandgrand) [inline components/base/ButtonLink/ButtonLink.test.js:25]: Same as SiteNavbar

File: components/base/ButtonLink/ButtonLink.test.js [hunk 10]
```diff
@@ -0,0 +1,58 @@
...
+    });
+    expect(getByText('Any%')).toBeInTheDocument();
+  });
+
+  it('renders with the passed link', () => {
+    stubbedRender(ButtonLink, {
+      attrs: defaultAttrs,
+      props: defaultProps,
+    });
+    expect(document.querySelector('a').getAttribute('href')).toBe(
```
Comment (legrandgrand) [inline components/base/ButtonLink/ButtonLink.test.js:42]: Again, would be nice to do selection with Testing Library, so it matches the idea that we test what the user see and not technical stuff

File: components/core/LoginButton/LoginButton.test.js [hunk 11]
```diff
@@ -0,0 +1,16 @@
...
+  it('should render without crashing', () => {
+    const { unmount } = stubbedRender(LoginButton);
+
+    unmount();
+  });
+
+  it('renders correctly', () => {
+    stubbedRender(LoginButton);
+
+    expect(document.querySelector('.login-button')).toMatchSnapshot();
```
Comment (legrandgrand) [inline components/core/LoginButton/LoginButton.test.js:14]: Same as SiteNavBar

File: components/core/NavLinks/NavLink/NavLink.test.js [hunk 12]
```diff
@@ -0,0 +1,27 @@
...
+  });
+
+  it('renders correctly', () => {
+    stubbedRender(NavLink, {
+      attrs: defaultAttrs,
+      props: defaultProps,
+      slots: { default: 'Games' },
+    });
+
+    expect(document.querySelector('.custom-link')).toMatchSnapshot();
```
Comment (legrandgrand) [inline components/core/NavLinks/NavLink/NavLink.test.js:25]: Same as SiteNavBar

File: components/core/NavLinks/NavLinks.test.js [hunk 13]
```diff
@@ -0,0 +1,38 @@
...
+
+    unmount();
+  });
+
+  it('renders correctly', () => {
+    stubbedRender(NavLinks, {
+      props: defaultProps,
+    });
+
+    expect(document.querySelector('.nav-link-container')).toMatchSnapshot();
```
Comment (legrandgrand) [inline components/core/NavLinks/NavLinks.test.js:26]: Same as SiteNavBar

File: components/core/NavLinks/NavLinks.test.js [hunk 14]
```diff
@@ -0,0 +1,38 @@
...
+  });
+
+  it('should render the same amount of <NavLink /> components as there are items in the navLinks props', () => {
+    stubbedRender(NavLinks, {
+      props: defaultProps,
+    });
+
+    expect(document.querySelectorAll('.nav-link').length).toEqual(
```
Comment (legrandgrand) [inline components/core/NavLinks/NavLinks.test.js:34]: Can you use testing library API, with maybe findAllByText or something like that?

File: components/core/SearchBar/SearchBar.test.js [hunk 15]
```diff
@@ -0,0 +1,47 @@
+import { fireEvent, stubbedRender } from '@/testUtils';
```
Comment (legrandgrand) [inline components/core/SearchBar/SearchBar.test.js:1]: SearchBar doesn't have 100% coverage, would be nice if we could reach that for this one too :)

File: components/core/SearchBar/SearchBar.test.js [hunk 16]
```diff
@@ -0,0 +1,47 @@
...
+
+    unmount();
+  });
+
+  it('renders correctly', () => {
+    stubbedRender(SearchBar, {
+      props: defaultProps,
+    });
+
+    expect(document.querySelector('.search-bar')).toMatchSnapshot();
```
Comment (legrandgrand) [inline components/core/SearchBar/SearchBar.test.js:21]: Same as SiteNavbar

File: components/core/SearchBar/SearchBar.test.js [hunk 17]
```diff
@@ -0,0 +1,47 @@
...
+  });
+
+  describe('when search is initiated', () => {
+    test('the onSubmit method is called when search button is clicked', async () => {
+      stubbedRender(SearchBar, {
+        props: defaultProps,
+      });
+
+      const searchButton = document.querySelector('button[type="submit"]');
```
Comment (legrandgrand) [inline components/core/SearchBar/SearchBar.test.js:30]: Can you use testing library to find the searchButton too? So we don't click on the first button of type submit but the button where it's written submit.

And also searchButton is only used in fireEvent, we can inline that for better readability

File: components/core/SearchBar/SearchBar.test.js [hunk 18]
```diff
@@ -0,0 +1,47 @@
...
+
+      expect(mockOnSubmit).toHaveBeenCalled();
+    });
+
+    test('the onSubmit method is called when the enter key is released', async () => {
+      stubbedRender(SearchBar, {
+        props: defaultProps,
+      });
+
+      const searchInput = document.querySelector('input');
```
Comment (legrandgrand) [inline components/core/SearchBar/SearchBar.test.js:41]: Same as above

File: components/core/SignUpButton/SignUpButton.test.js [hunk 19]
```diff
@@ -0,0 +1,16 @@
...
+  it('should render without crashing', () => {
+    const { unmount } = stubbedRender(SignUpButton);
+
+    unmount();
+  });
+
+  it('renders correctly', () => {
+    stubbedRender(SignUpButton);
+
+    expect(document.querySelector('.sign-up-button')).toMatchSnapshot();
```
Comment (legrandgrand) [inline components/core/SignUpButton/SignUpButton.test.js:14]: Same as SiteNavbar

**Review by legrandgrand — CHANGES_REQUESTED** [review]
Very good PR! Lots of cool changes in there: functional husky, near 100% with unit test coverage, CI, lodash is nice, lots of cool scripts to make our life better.

However a few changes should be done before merging:
- Have 100% coverage (we are really close to that
- Test using Testing Library API instead of using document selectors
- Run prettier once on the whole project now that it works with husky (though that's debatable I guess... Will probably bite us later)
- Add missing dependencies that are in package-lock but not package.json
- Only use lodash when the function doesn't exist directly nativly (join, forEach...)

One last thing I didn't mention:
all the tests are in .js files and therefore not typed. Should we move on to TS with this directly or create another issue?

File: components/core/SearchBar/SearchBar.test.js (hunk 17 above)
Comment (erunks) [reply components/core/SearchBar/SearchBar.test.js]: I went with the `data-testid` attribute here, in order to make sure we're selecting the correct elements, without effecting anything in the DOM.

**Comment by erunks** [comment]
> One last thing I didn't mention:
> all the tests are in .js files and therefore not typed. Should we move on to TS with this directly or create another issue?

So specifically around this, I think it's probably best to make another issue for it. I tried to address it with a commit in this PR, but I was having a bunch of trouble with it, especially since I've never setup anything TS related on my own. So I'd love to have someone more familiar with TS help out with this.

**Review by legrandgrand — APPROVED** [review]

**Review by maxcaplan — APPROVED** [review]
Looks good to me!

**Pull request merged**